### ISSUE TITLE
Add timeout to XMPP IQ requests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2218]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2218|
|Type|enhancement|
|Created|2006-11-04 18:14:26Z|

XMPP <iq/> stanzas are designed to have request-response behaviour. The current implementation of t.w.p.j.xmlstream.IQ has a send() method that returns a deferred which fires when a response has been received. 

However, if for some reason the other party is too slow or unable to respond, it is desirable to have those deferreds fire with a failure that indicates no response has been received within a configured time period.

Note that the current implementation already fires the deferreds of all pending requests when the XML stream connection has been lost.

<details><summary>Searchable metadata</summary>

```
trac-id__2218 2218
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_5 Words-0.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1162664066000000 1162664066000000
changetime__1163338643000000 1163338643000000
version__None None
owner__ralphm ralphm
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Implemented in [branches/xmpp-iq-timeout-2218.](branches/xmpp-iq-timeout-2218.)

To test timeouts, I used t.i.task.Clock, and uncovered that it doesn't set its called attribute properly, as mentioned in [#2222](https://github.com/twisted/twisted/issues/2222). This branch contains a fix for this.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @ralphm<br>@exarkun set status to `new`|
|-|-|

Fix to task.py looks good.  Timeout support added to jabber mostly looks good.  I wonder what will happen if a response is received after the timeout expires, though.  Add a test for this case?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set owner to @exarkun|
|-|-|

Nothing happens, since the iq's id is removed from the `iqDeferreds` dictionary. The existing tests already test this. I did add a test for making sure untracked iq responses (id not in that dictionary) are passed untouched to the next observer.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @ralphm|
|-|-|

Looks good.  Go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm set status to `closed`|
|-|-|

(In [18644]) Add timeout behaviour to xmpp iq requests and make t.i.Clock set call.called.

Author: ralphm
Reviewer: exarkun
Fixes [#2218](https://github.com/twisted/twisted/issues/2218) and [#2222](https://github.com/twisted/twisted/issues/2222)
